x/auth/middleware/run_msgs.go

```diff
@@ -0,0 +1,166 @@
+	res, err := txh.runMsgs(sdk.UnwrapSDKContext(ctx), tx.GetMsgs(), req.Tx)
+	if err != nil {
+		return abci.ResponseDeliverTx{}, err
+	}
+
+	return abci.ResponseDeliverTx{
+		// GasInfo will be populated by the antehandlers.
+		Log:    res.Log,
+		Data:   res.Data,
+		Events: res.Events,
+	}, nil
+}
+
+// SimulateTx implements tx.Handler.SimulateTx method.
+func (txh runMsgsTxHandler) SimulateTx(ctx context.Context, sdkTx sdk.Tx, req tx.RequestSimulateTx) (tx.ResponseSimulateTx, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	res, err := txh.runMsgs(sdkCtx, sdkTx.GetMsgs(), req.TxBytes)
+	if err != nil {
+		return tx.ResponseSimulateTx{}, err
+	}
+
+	return tx.ResponseSimulateTx{
+		// GasInfo will be populated by the antehandlers.
+		Result: res,
+	}, nil
+}
+
+// runMsgs iterates through a list of messages and executes them with the provided
+// Context and execution mode. Messages will only be executed during simulation
+// and DeliverTx. An error is returned if any single message fails or if a
+// Handler does not exist for a given message route. Otherwise, a reference to a
+// Result is returned. The caller must not commit state if an error is returned.
+func (txh runMsgsTxHandler) runMsgs(sdkCtx sdk.Context, msgs []sdk.Msg, txBytes []byte) (*sdk.Result, error) {
+	// Create a new Context based off of the existing Context with a MultiStore branch
+	// in case message processing fails. At this point, the MultiStore
+	// is a branch of a branch.
+	runMsgCtx, msCache := cacheTxContext(sdkCtx, txBytes)
+
+	// Attempt to execute all messages and only update state if all messages pass
+	// and we're in DeliverTx. Note, runMsgs will never return a reference to a
+	// Result if any single message fails or does not have a registered Handler.
+	msgLogs := make(sdk.ABCIMessageLogs, 0, len(msgs))
+	events := sdkCtx.EventManager().Events()
+	txMsgData := &sdk.TxMsgData{
+		Data: make([]*sdk.MsgData, 0, len(msgs)),
+	}
+
+	// NOTE: GasWanted is determined by the AnteHandler and GasUsed by the GasMeter.
+	for i, msg := range msgs {
+		var (
+			msgResult    *sdk.Result
+			eventMsgName string // name to use as value in event `message.action`
+			err          error
+		)
+
+		if handler := txh.msgServiceRouter.Handler(msg); handler != nil {
+			// ADR 031 request type routing
+			msgResult, err = handler(runMsgCtx, msg)
+			eventMsgName = sdk.MsgTypeURL(msg)
+		} else if legacyMsg, ok := msg.(legacytx.LegacyMsg); ok {
+			// legacy sdk.Msg routing
+			// Assuming that the app developer has migrated all their Msgs to
+			// proto messages and has registered all `Msg services`, then this
+			// path should never be called, because all those Msgs should be
+			// registered within the `MsgServiceRouter` already.
+			msgRoute := legacyMsg.Route()
+			eventMsgName = legacyMsg.Type()
+			handler := txh.legacyRouter.Route(sdkCtx, msgRoute)
```

Comment:
Are we going to deprecate the legacy router?

Reply:
We deprecated it already. But we didn't remove it, it makes more sense to do it when we remove Amino msgs.